components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/dao/impl/IdentityUserMetadataMgtDAOImpl.java

```diff
@@ -0,0 +1,508 @@
+import org.wso2.carbon.identity.event.IdentityEventException;
+import org.wso2.carbon.identity.governance.internal.cache.IdentityDataStoreCache;
+import org.wso2.carbon.identity.governance.internal.cache.IdentityDataStoreCacheKey;
+import org.wso2.carbon.identity.governance.model.UserIdentityClaim;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryConstants;
+import org.wso2.carbon.identity.recovery.dao.IdentityUserMetadataMgtDAO;
+import org.wso2.carbon.identity.recovery.internal.IdentityRecoveryServiceDataHolder;
+import org.wso2.carbon.user.api.RealmConfiguration;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.claim.ClaimManager;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.user.core.util.DatabaseUtil;
+import org.wso2.carbon.user.core.util.UserCoreUtil;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Locale;
+import java.util.Map;
+import javax.sql.DataSource;
+
+/**
+ * Implementation class for IdentityUserMetadataMgtDAO.
+ */
+public class IdentityUserMetadataMgtDAOImpl implements IdentityUserMetadataMgtDAO {
+
+    private static final Log log = LogFactory.getLog(IdentityUserMetadataMgtDAOImpl.class);
+
+    private final IdentityDataStoreCache identityDataStoreCache = IdentityDataStoreCache.getInstance();
+
+    @Override
+    public void updateUserMetadata(UserStoreManager userStoreManager, String username,
+                                   String claimURI, String value, String eventName) throws IdentityEventException {
+
+        if (username.contains(IdentityRecoveryConstants.TENANT_ASSOCIATION_MANAGER)) {
+            return;
+        }
+
+        String tenantId;
+        try {
+            tenantId = String.valueOf(userStoreManager.getTenantId());
+        } catch (UserStoreException e) {
+            throw new IdentityEventException(
+                    String.format("Error occurred while retrieving the tenantId of the user related to %s event.",
+                            eventName), e);
+        }
+
+        String userStoreDomain = UserCoreUtil.extractDomainFromName(username);
+        if (StringUtils.isNotBlank(userStoreDomain) && IdentityRecoveryConstants.ASGARDEO_USER_DOMAIN_NAME
+                .equalsIgnoreCase(userStoreDomain) && isStoreIdentityClaimsInUserStoreEnabled(userStoreManager)) {
+            username = UserCoreUtil.removeDomainFromName(username);
+            storeMetadataToUserStore(userStoreManager, username, tenantId, claimURI, value, eventName);
+            return;
+        }
+
+        storeMetadataToIdentityStore(username, tenantId, claimURI, value, eventName);
+    }
+
+    @Override
+    public UserIdentityClaim loadUserMetadataFromCache(UserStoreManager userStoreManager, String userName) {
+
+        try {
+            if (userName != null) {
```

Comment:
Can we negate this condition? If so we can avoid shifting the code to the right.